core/src/serde/de.rs

```diff
@@ -0,0 +1,693 @@
+        _name: &str,
+        _fields: &[&str],
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        match self {
+            Self::Map(map) => visit_map(map, visitor),
+            _ => error(format!(
+                "Only `Ipld::Map` can be deserialized to struct, input was `{:#?}`",
+                self
+            )),
+        }
+    }
+
+    fn deserialize_unit_struct<V: de::Visitor<'de>>(
+        self,
+        _name: &str,
+        _visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        error("Unit struct cannot be deserialized")
+    }
+
+    fn deserialize_newtype_struct<V: de::Visitor<'de>>(
+        self,
+        name: &str,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        if name == CID_SERDE_PRIVATE_IDENTIFIER {
+            match self {
+                Ipld::Link(cid) => visitor.visit_newtype_struct(CidDeserializer(cid)),
+                _ => error(format!(
+                    "Only `Ipld::Link`s can be deserialized to CIDs, input was `{:#?}`",
+                    self
+                )),
+            }
+        } else {
+            visitor.visit_newtype_struct(self)
+        }
+    }
+
+    // Heavily based on
+    // https://github.com/serde-rs/json/blob/95f67a09399d546d9ecadeb747a845a77ff309b2/src/value/de.rs#L249
+    fn deserialize_enum<V: de::Visitor<'de>>(
+        self,
+        _name: &str,
+        _variants: &[&str],
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        let (variant, value) = match self {
+            Ipld::Map(map) => {
+                let mut iter = map.into_iter();
+                let (variant, value) = match iter.next() {
+                    Some(v) => v,
+                    None => {
+                        return error(
+                            "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had no keys"
+                        );
+                    }
+                };
+                // Enums are encoded in IPLD as maps with a single key-value pair
+                if iter.next().is_some() {
+                    return error(
+                        "Only `Ipld::Map`s with a single key can be deserialized to `enum`, input had more keys"
+                    );
+                }
+                (variant, Some(value))
+            }
+            Ipld::String(variant) => (variant, None),
+            _ => return error(format!(
+                    "Only `Ipld::Map` and `Ipld::String` can be deserialized to `enum`, input was `{:#?}`",
+                    self
+            )),
+        };
+
+        visitor.visit_enum(EnumDeserializer { variant, value })
+    }
+
+    // Heavily based on
+    // https://github.com/serde-rs/json/blob/95f67a09399d546d9ecadeb747a845a77ff309b2/src/value/de.rs#L446
+    fn deserialize_ignored_any<V: de::Visitor<'de>>(
+        self,
+        visitor: V,
+    ) -> Result<V::Value, Self::Error> {
+        drop(self);
+        visitor.visit_none()
```

Comment:
You know, I wonder if this should be a `visit_unit`?

Reply:
Oh, I see. It literally doesn't matter. `deserialize_ignored_any` is just a hint to the deserializer that it shouldn't _actually_ bother decoding anything. The visitor will "go through the motions", but will ignore any results. E.g., if you ask it to visit a map, it'll do that then throw away the results.

So calling `visit_something_free` is the right way here (none, unit, etc.). Let's leave it as-is.

Reply:
I'd leave it as is.